### PR TITLE
fix(model): apply final logit softcapping to Gemma 4 dense models

### LINKED ISSUE #4610: [bug] Gemma4 Dense model lacks final logit softcapping
### Problem

In the [request](https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4148) final logit softcapping was added only for MoE models

### Minimal repro

```shell
Run Gemma4
```

### Expected behavior

Logits are capped

### Affected area

area:model

### Regression?

Yes

### Environment

_No response_

### Logs

```shell

```

### PR DESCRIPTION
**Pull request by cuichenx**
## What changed

Gemma 4 dense models now preserve Hugging Face's `final_logit_softcapping` value and apply exactly one final `scale * tanh(logits / scale)` operation after the LM head. The behavior is shared by the flat text and nested VL providers and exports, and installation is idempotent so an inherited/custom output layer cannot be capped twice.

This is the focused fix for the bug reported in #4610. The independently discovered E2B heterogeneous-FFN and 26B-A4B MoE conversion/runtime corrections are split into a separate follow-up PR.

Fixes #4610

## Root cause

PR #4148 added the existing `Gemma4OutputLayer` softcap only to the MoE provider. The dense provider dropped the config field and never installed that output behavior. Weight roundtrips could still be exact because softcapping is a config/runtime operation, and greedy IDs could still match because positive-scale `tanh` is monotonic and preserves argmax ordering.

## Validation

- `nvcr.io/nvidia/nemo:26.06`, Python 3.12, Transformers 5.8.1
- focused bridge/provider/VL unit tests: `213 passed`
- controlled logits drive values well beyond the cap and distinguish zero, one, and two applications
- disabled/missing caps, dense text/VL config propagation and export, and idempotent dense/MoE construction are covered
- `uv run pre-commit run --all-files`

Pinned official checkpoints used for the original reproduction and post-fix controlled parity:

- `google/gemma-4-E2B-it@70af34e20bd4b7a91f0de6b22675850c43922a03`
- `google/gemma-4-E4B-it@fee6332c1abaafb77f6f9624236c63aa2f1d0187`
- `google/gemma-4-26B-A4B-it@20da991ab4afab98e8f910c4a2e8f4fbefc404ad`
- `google/gemma-4-31B-it@3548789868c5356dbf307c98e6f609007b82b3eb`

For the 8x output-head saturation control, dense post-softcap HF-vs-MCore max/mean absolute errors changed from E2B `1114/624.874` to `0/0` (real head isolation), E4B `586/253.087` to `2.5625/0.000411`, and 31B `430/103.032` to `2.4375/0.006614`. E4B and 31B full-model greedy outputs remained exactly equal. The existing 26B-A4B MoE path still applies a single cap: its runtime is within max `0.122` of a single cap of its own raw logits and is clearly separated from a double cap.

The residual dense maxima are BF16 fused-GEMM accumulation differences: layerwise comparison localizes them before the output operation, exact-head isolation removes them, and an FP32 E4B control reduces final raw-logit max error to `1.221e-4`.

### CODE REVIEW

**Review by yaoyu-33 — APPROVED** [review]

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4645/commits/a6e1e967f3ff40e62088ca58b1321ee2d350b9ea

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4645/commits/a6e1e967f3ff40e62088ca58b1321ee2d350b9ea

**Comment by cuichenx** [comment]
/ok to test https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/4645/commits/38899ba6b677453ca40f6904726f9a3d3ef63ca4
